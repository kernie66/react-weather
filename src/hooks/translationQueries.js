import { useQueryClient } from '@tanstack/react-query';
import axios from 'axios';
import { isEmpty } from 'radash';
import queryPersister from '../helpers/queryPersister.js';
import { useCallback } from 'react';

const googleApiKey = import.meta.env.VITE_GOOGLEMAPS_API_KEY;
const googleApiUrl =
  'https://translation.googleapis.com/language/translate/v2';
const fullApiUrl = `${googleApiUrl}?key=${googleApiKey}`;

const getTranslation = async ({
  queryKey: [, { texts, language }], // Get the objects of queryKey[1]
}) => {
  if (isEmpty(texts)) {
    console.log('Error: No text supplied to translate');
    throw new Error('No text supplied to translate');
  }

  const { data } = await axios.post(fullApiUrl, {
    q: texts,
    target: language,
  });
  return data.data.translations; // return array of translations
};

export const useTranslation = (defaultLanguage = 'sv') => {
  const queryClient = useQueryClient();

  const translate = useCallback(
    async (text, language = defaultLanguage) => {
      console.log('Hook text to translate:', text);
      const translatedTexts = await queryClient.fetchQuery({
        queryKey: [
          'translations',
          { texts: text, language: language },
        ],
        queryFn: getTranslation,
        staleTime: Infinity,
        gcTime: 1000 * 60 * 60 * 24 * 7, // One week
        persister: queryPersister(),
      });
      // setTextToTranslate(text);
      // setTargetLanguage(language);
      console.log('Hook translatedTexts:', translatedTexts);
      return translatedTexts;
    },
    [defaultLanguage, queryClient]
  );

  return translate;
};
