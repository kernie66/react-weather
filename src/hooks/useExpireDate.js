import dayjs from 'dayjs';
import { useEffect, useState } from 'react';

const useExpireDate = ({
  expireHour,
  expireMinute = 0,
  expireSecond = 0,
}) => {
  const [expired, setExpired] = useState(false);
  const [timeoutTime, setTimeoutTime] = useState();

  useEffect(() => {
    const now = dayjs();
    console.log('ðŸš€ ~ useExpireDate ~ now:', now);
    let setExpireHour = expireHour % 24 || now.hour();
    let setExpireMinute = expireMinute % 60;
    let setExpireSecond = expireSecond % 60;

    console.log('ðŸš€ ~ now:', now.hour());
    // Check if time already passed
    if (now.hour() >= expireHour) {
      setExpireHour += 24;
    }

    console.log('ðŸš€ ~ setExpireMinute:', setExpireMinute);

    const expireDate = now
      .hour(setExpireHour)
      .minute(setExpireMinute)
      .second(setExpireSecond);
    console.log('ðŸš€ ~ useExpireDate ~ expireDate:', expireDate);
    setTimeoutTime(expireDate.diff(now));
  }, [expireHour, expireMinute, expireSecond]);

  useEffect(() => {
    if (timeoutTime) {
      console.log('ðŸš€ ~ useExpireDate ~ timeoutTime:', timeoutTime);
      const timeoutRef = setTimeout(() => {
        setExpired(true);
      }, timeoutTime);

      return () => {
        console.log('Clear timeout');
        clearTimeout(timeoutRef);
      };
    }
  }, [timeoutTime]);

  return expired;
};

export default useExpireDate;
