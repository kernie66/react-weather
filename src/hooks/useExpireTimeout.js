import dayjs from 'dayjs';
import { useEffect, useState } from 'react';

export const getExpireTime = ({
  expireHour,
  expireMinute,
  expireSecond,
  relativeTime = false,
}) => {
  const now = dayjs();
  console.log('ðŸš€ ~ useExpireDate ~ now:', now);
  let setExpireHour = expireHour % 24 || now.hour();
  let setExpireMinute = expireMinute % 60 || 0;
  let setExpireSecond = expireSecond % 60 || 0;

  console.log('ðŸš€ ~ now:', now.hour());
  // Check if time already passed
  if (now.hour() >= expireHour) {
    setExpireHour += 24;
  } else if (now.minute() >= expireMinute) {
    setExpireMinute += 1;
  }

  console.log('ðŸš€ ~ setExpireMinute:', setExpireMinute);

  const expireDate = now
    .hour(setExpireHour)
    .minute(setExpireMinute)
    .second(setExpireSecond);
  console.log('ðŸš€ ~ useExpireDate ~ expireDate:', expireDate);
  const expireTime = expireDate.diff(now);
  console.log('ðŸš€ ~ expireTime:', expireTime);

  return expireTime;
};

export const useExpireTimeout = ({ expireTime = 1000 }) => {
  const [expired, setExpired] = useState(false);

  useEffect(() => {
    if (expireTime) {
      console.log('ðŸš€ ~ useExpireDate ~ timeoutTime:', expireTime);
      const timeoutRef = setTimeout(() => {
        setExpired(true);
      }, expireTime);

      return () => {
        console.log('Clear timeout');
        clearTimeout(timeoutRef);
      };
    }
  }, [expireTime]);

  return expired;
};
