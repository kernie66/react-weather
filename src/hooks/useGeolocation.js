import { useEffect, useState } from 'react';
import { defaultPosition } from '../atoms/locationStates.js';

export const useGeolocation = () => {
  const [latitude, setLatitude] = useState(defaultPosition.lat);
  const [longitude, setLongitude] = useState(defaultPosition.lng);

  const handleSuccess = ({ coords: { latitude, longitude } }) => {
    setLatitude(latitude);
    setLongitude(longitude);
  };

  useEffect(() => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(handleSuccess);
    } else {
      console.log('Geolocation not available');
    }
  }, []);

  return {
    latitude,
    longitude,
  };
};
