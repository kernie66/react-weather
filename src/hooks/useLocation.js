import { useContext } from 'react';
import { LocationContext } from '../contexts/Contexts.js';

const defaultLat = import.meta.env.VITE_DEFAULT_LATITUDE;
const defaultLon = import.meta.env.VITE_DEFAULT_LONGITUDE;
export const defaultPosition = {
  lat: parseFloat(defaultLat),
  lng: parseFloat(defaultLon),
};
export const defaultAddress = 'Rotebro, Sollentuna, Sverige';

export default function useLocation() {
  return useContext(LocationContext);
}
