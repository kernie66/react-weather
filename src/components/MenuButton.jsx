import { TbRefresh } from 'react-icons/tb';
import { TbMenu2 } from 'react-icons/tb';
import { FaCompress, FaExpand } from 'react-icons/fa';
import {
  Button,
  Menu,
  rem,
  Text,
  VisuallyHidden,
} from '@mantine/core';
import { useFullscreen } from '@mantine/hooks';

export default function MenuButton() {
  const { toggle, fullscreen } = useFullscreen();

  const reload = () => {
    window.location.reload();
  };

  return (
    <Menu position="bottom-start">
      <Menu.Target>
        <Button px={0} variant="subtle" fz={36} c="teal">
          <TbMenu2 />
          <VisuallyHidden>Meny</VisuallyHidden>
        </Button>
      </Menu.Target>
      <Menu.Dropdown>
        <Menu.Item
          onClick={reload}
          leftSection={
            <TbRefresh
              color="teal"
              style={{ width: rem(20), height: rem(20) }}
            />
          }
        >
          <Text fz={20}>Uppdatera</Text>
        </Menu.Item>
        <Menu.Item
          onClick={toggle}
          leftSection={
            fullscreen ? (
              <FaCompress
                color="teal"
                style={{ width: rem(20), height: rem(20) }}
              />
            ) : (
              <FaExpand
                color="teal"
                style={{ width: rem(20), height: rem(20) }}
              />
            )
          }
        >
          <Text fz={20}>{fullscreen ? 'Återställ' : 'Helskärm'}</Text>
        </Menu.Item>
      </Menu.Dropdown>
    </Menu>
  );
}
