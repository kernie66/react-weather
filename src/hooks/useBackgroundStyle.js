import { useAtomValue } from 'jotai';
import { backgroundImageState } from '../atoms/weatherThemeStates.js';
import { getWeatherImageUrl } from '../helpers/getImageUrl.js';
import { useImageSize } from 'react-image-size';
import { useViewportSize } from '@mantine/hooks';
import { getBackgroundSizeStyle } from '../helpers/getBackgroundSizeStyle.js';

export const useBackgroundStyle = () => {
  const backgroundImage = useAtomValue(backgroundImageState);

  const backgroundImageUrl = getWeatherImageUrl(
    `${backgroundImage}.jpg`
  );
  const [dimensions] = useImageSize(backgroundImageUrl);
  const screenSize = useViewportSize();
  const backgroundSize = getBackgroundSizeStyle(
    dimensions,
    screenSize
  );
  const backgroundStyle = {
    backgroundImage: `url(${backgroundImageUrl})`,
    backgroundPosition: 'center',
    backgroundSize: backgroundSize,
    backgroundRepeat: 'no-repeat',
    // width: '100vw',
    height: '100vh',
  };
  return backgroundStyle;
};
