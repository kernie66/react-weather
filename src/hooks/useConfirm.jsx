// import PropTypes from "prop-types";
import { useState } from 'react';
import createPromise from '../helpers/createPromise';
import { useDisclosure } from '@mantine/hooks';
import { Button, Divider, Group, Modal } from '@mantine/core';

export const useConfirm = () => {
  const [opened, { open, close }] = useDisclosure(false);
  const [resolver, setResolver] = useState({ resolver: null });

  const getConfirmation = () => {
    open();
    const [promise, resolve] = createPromise();
    setResolver({ resolve });
    return promise;
  };

  const onClose = async (status) => {
    close();
    resolver.resolve(status);
  };

  const ConfirmModal = ({
    message,
    title,
    confirmText,
    cancelText,
    confirmColor,
    cancelColor,
    size,
  }) => {
    return (
      <Modal
        opened={opened}
        onClose={() => onClose(false)}
        size={size}
        centered
        title={title}
      >
        <Divider mb={8} />
        {message}
        <Group justify="space-between" my={8} pt={16}>
          <Button
            color={confirmColor}
            onClick={() => onClose(true)}
            data-autofocus
          >
            {confirmText}
          </Button>
          <Button
            color={cancelColor}
            variant="outline"
            onClick={() => onClose(false)}
          >
            {cancelText}
          </Button>
        </Group>
      </Modal>
    );
  };

  return [getConfirmation, ConfirmModal];
};
