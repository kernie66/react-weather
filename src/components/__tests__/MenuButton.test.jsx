import { describe, vi, expect } from 'vitest';
import MenuButton from '../MenuButton.jsx';
import { render, screen } from '../../../testing-utils';
import { userEvent } from '../../../testing-utils/index.js';
import { useFullscreen } from '@mantine/hooks';

const toggleSpy = vi.fn();

vi.mock('@mantine/hooks');

describe('test MenuButton', () => {
  const original = window.location;

  beforeAll(() => {
    Object.defineProperty(window, 'location', {
      configurable: true,
      value: { reload: vi.fn() },
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  afterAll(() => {
    Object.defineProperty(window, 'location', {
      configurable: true,
      value: original,
    });
    vi.restoreAllMocks();
  });

  it('renders menu button and opens the menu', async () => {
    const user = userEvent.setup();

    useFullscreen.mockReturnValue({
      toggle: toggleSpy,
      fullscreen: false,
    });

    render(<MenuButton />);

    const menuButton = screen.getByRole('button', {
      name: /meny/i,
    });
    expect(menuButton).toBeInTheDocument();
    expect(useFullscreen).toHaveBeenCalledOnce();

    // Open the menu
    await user.click(menuButton);

    // Check that the menu items are rendered
    expect(
      await screen.findByRole('menuitem', {
        name: /uppdatera/i,
      })
    ).toBeInTheDocument();
    expect(
      await screen.findByRole('menuitem', {
        name: /helsk채rm/i,
      })
    ).toBeInTheDocument();
    expect(toggleSpy).not.toHaveBeenCalled();
  });

  it('selects update from the menu', async () => {
    const user = userEvent.setup();

    useFullscreen.mockReturnValue({
      toggle: toggleSpy,
      fullscreen: false,
    });

    render(<MenuButton />);

    const menuButton = screen.getByRole('button', {
      name: /meny/i,
    });
    expect(menuButton).toBeInTheDocument();
    expect(vi.isMockFunction(window.location.reload)).toBe(true);
    expect(useFullscreen).toHaveBeenCalledOnce();

    // Open the menu
    await user.click(menuButton);

    // Get the menu item update
    const menuItemUpdate = await screen.findByRole('menuitem', {
      name: /uppdatera/i,
    });
    expect(menuItemUpdate).toBeInTheDocument();

    // Check that reload is called
    await user.click(menuItemUpdate);
    expect(menuItemUpdate).toBeInTheDocument();
    expect(window.location.reload).toHaveBeenCalled();
    expect(toggleSpy).not.toHaveBeenCalled();
  });

  it('in normal screen, selects fullscreen from the menu', async () => {
    const user = userEvent.setup();

    useFullscreen.mockReturnValue({
      toggle: toggleSpy,
      fullscreen: false,
    });

    render(<MenuButton />);

    const menuButton = screen.getByRole('button', {
      name: /meny/i,
    });
    expect(menuButton).toBeInTheDocument();
    expect(useFullscreen).toHaveBeenCalledOnce();
    expect(useFullscreen).toHaveReturnedWith(
      expect.objectContaining({
        fullscreen: false,
      })
    );

    // Open the menu
    await user.click(menuButton);

    // Get the fullscreen menu item
    const menuItemFullScreen = await screen.findByRole('menuitem', {
      name: /helsk채rm/i,
    });
    expect(menuItemFullScreen).toBeInTheDocument();
    expect(toggleSpy).not.toHaveBeenCalled();

    // Click on fullscreen
    await user.click(menuItemFullScreen);
    expect(menuButton).toBeInTheDocument();
    expect(toggleSpy).toHaveBeenCalledTimes(1);
  });

  it('in fullscreen, selects normal screen from the menu', async () => {
    const user = userEvent.setup();

    useFullscreen.mockReturnValue({
      toggle: toggleSpy,
      fullscreen: true,
    });

    render(<MenuButton />);

    const menuButton = screen.getByRole('button', {
      name: /meny/i,
    });
    expect(menuButton).toBeInTheDocument();
    expect(useFullscreen).toHaveBeenCalledOnce();
    expect(useFullscreen).toHaveReturnedWith(
      expect.objectContaining({
        fullscreen: true,
      })
    );

    // Open the menu
    await user.click(menuButton);

    // Get the fullscreen menu item
    const menuItemRestore = await screen.findByRole('menuitem', {
      name: /책terst채ll/i,
    });
    expect(menuItemRestore).toBeInTheDocument();
    expect(toggleSpy).not.toHaveBeenCalled();

    // Click on fullscreen
    await user.click(menuItemRestore);
    expect(menuButton).toBeInTheDocument();
    expect(toggleSpy).toHaveBeenCalledTimes(1);
  });
});
