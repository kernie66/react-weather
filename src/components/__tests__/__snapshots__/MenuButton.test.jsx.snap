// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test MenuButton > renders menu button and opens the menu 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <style
        data-mantine-styles="classes"
      >
        @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
      </style>
      <button
        aria-controls=""
        aria-expanded="false"
        aria-haspopup="menu"
        class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
        data-variant="subtle"
        id=""
        style="--button-bg: transparent; --button-hover: var(--mantine-color-blue-light-hover); --button-color: var(--mantine-color-blue-light-color); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; padding-inline: 0rem;"
        type="button"
      >
        <span
          class="m_80f1301b mantine-Button-inner"
        >
          <span
            class="m_811560b9 mantine-Button-label"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6l16 0"
              />
              <path
                d="M4 12l16 0"
              />
              <path
                d="M4 18l16 0"
              />
            </svg>
            <span
              class="m_515a97f8 mantine-VisuallyHidden-root"
            >
              Meny
            </span>
          </span>
        </span>
      </button>
    </div>
    <div
      data-portal="true"
    />
  </body>,
  "container": <div>
    <style
      data-mantine-styles="classes"
    >
      @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
    </style>
    <button
      aria-controls=""
      aria-expanded="false"
      aria-haspopup="menu"
      class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
      data-variant="subtle"
      id=""
      style="--button-bg: transparent; --button-hover: var(--mantine-color-blue-light-hover); --button-color: var(--mantine-color-blue-light-color); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; padding-inline: 0rem;"
      type="button"
    >
      <span
        class="m_80f1301b mantine-Button-inner"
      >
        <span
          class="m_811560b9 mantine-Button-label"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6l16 0"
            />
            <path
              d="M4 12l16 0"
            />
            <path
              d="M4 18l16 0"
            />
          </svg>
          <span
            class="m_515a97f8 mantine-VisuallyHidden-root"
          >
            Meny
          </span>
        </span>
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
